dg-commons/src/main/code/Graph/Graph.scala

```diff
@@ -0,0 +1,210 @@
+package Graph
+
+import java.io.{FileOutputStream, OutputStream}
+
+import Helpers.{DotHelper, RandomHelper}
+import NodeData.DisplayableData
+
+import scala.collection.mutable
+
+/** *
+  * Generic directed acyclic graph. Can have multiple root nodes.
+  * Programmers should be careful not to create cycles, because this class does not check for cycles.
+  * @tparam T Type of data stored in each graph childNode
+  * @param initialNodeValue Create the graph with one childNode containing this supplied value.
+  * @param isEdgeLinkTrackingOn Whether or not to maintain history of all events used to create graph.
+  * @param _graphId Graph ID. Can be set later using graphId setter method. If not set, will eventually default to UUID.
+  * @param appendSharedDisplayIdsWithNumericalSuffix If true, if multiple nodes with same displayableDataId, appends with _1, _2, etc.
+  */
+class Graph[T <: DisplayableData](initialNodeValue: Option[T] = None, var isEdgeLinkTrackingOn: Boolean = false, private var _graphId: String = "", var appendSharedDisplayIdsWithNumericalSuffix: Boolean = false) {
+  def this(graphId: String) {
+    this(_graphId=graphId, appendSharedDisplayIdsWithNumericalSuffix = true)
+  }
+  def this(initialNodeValue: T, graphId: String) {
+    this(initialNodeValue=Some(initialNodeValue), _graphId=graphId)
+  }
+
+  var customSeed: Option[Long] = None
+  var customGlobalSeed: Option[Short] = None
+  val customBooleanAttributes = mutable.HashMap[String, Boolean]()
+  var userConfiguredGraphId = false
+
+  /**
+   * Get the graph ID.
+   * @return
+   */
+  def graphId: String = {
+    if (_graphId.isEmpty) _graphId = RandomHelper.randomUuid().replace('-','_')
+    _graphId
+  }
+
+  /**
+   * Set the graph ID.
+   * @param value String to set as graph ID
+   */
+  def graphId_=(value: String) = {
+    _graphId = value.replace('-','_').replace(' ','_').replace("\"","")
+    userConfiguredGraphId = true
+  }
+
+  if (_graphId != "") graphId = _graphId
+
```

Comment:
This is good feedback.

This is one of the unfortunate things about using Scala code from Java. In Scala, this is equivalent to "var graphId: String" with some custom logic when getting and setting the value. The way you would use def graphId: String, in Scala, is "graphId" on the right-hand side of an expression, and the way you would use def graphId_(value: String) in Scala is "graphId = "some string"

It's actually very clean in Scala. The ugly part is when you want to use it from Java, and then it becomes:
graphId() for the getter -- not too bad not the parentheses do make it a bit worse.
graphId_=(value: String), instead of just saying `graphId = "some value"` as in Scala, becomes `graphId_$eq("some value")`, which is really ugly.

Maybe I can just add overrides such as "def setGraphId(value: String): Unit = { graphId = value }" for use in Java. And for any normal vars in Scala I can probably annotate them with @BeanProperty to generate the default Java getter/setter.